E2E test node upgrade (to same version)

mbforbes opened the pull request:
For GCE, do a rolling-update on the node managed instance group to the template that it's already using. This tests the node upgrade mechanism itself, separately from cross-version testing. Tests #6088; starts addressing #7914 (though not across multiple versions).

This is named the 'restart' e2e test, and is structured so that it can be implemented by other providers.

This also cleans up a few things:
- refactors and renames a couple constants to be consistent across tests
- formats some logging output so that big streams can be quickly eyeballed
- makes the reboot test more robust by retrying single API calls
- refactors a few helper functions for use across reboot and restart tests

mbforbes committed:
E2E test node upgrade (to same version)

mbforbes commented:
+cc @zmerlynn (review me maybe?)
+cc @roberthbailey (review me maybe?)
+cc @quinton-hoole (review me maybe?)
+cc @ixdy (review me maybe?)

quinton-hoole commented:
I can take a look.

roberthbailey commented:
Thanks @quinton-hoole. :)

jayunit100 commented on test/e2e/reboot.go:
```diff
@@ -126,8 +115,7 @@ func rebootNode(c *client.Client, provider, name string, result chan bool) {
 	// Get all the pods on the node.
-	podList, err := c.Pods(api.NamespaceDefault).List(
-		labels.Everything(), fields.OneTermEqualSelector(client.PodHost, name))
+	podList, err := listPods(c, api.NamespaceDefault, labels.Everything(), fields.OneTermEqualSelector(client.PodHost, name))
```
Uhoh... Looks like this test is using the default namespace. we can fix this now or later - if later - lets create an issue before commiting so we dont forget :)

quinton-hoole replied:
May as well do this now, before merge.
You can use @lavalamp 's new framework as per example in #8747
f := NewFramework("reboot")

mbforbes replied:
Ack, will do this.

mbforbes replied:
Done (not yet pushed).

mbforbes replied:
Agh, I should have looked into what this did before switching to it.

We don't want to create our own namespace for this test. This test is supposed to look at all of the infrastructure pods (in the default namespace) and ensure that they are running both before and after the cluster is restarted. While other tests should be isolated by namespace, this test is lower level than most tests,.

quinton-hoole commented on test/e2e/restart.go:
```diff
@@ -0,0 +1,367 @@
+	// How long each node is given during a process that restarts all nodes
+	// before the test is considered failed. (Note that the total time to
+	// restart all nodes will be this number times the nubmer of nodes.)
+	restartPerNodeTimeout = 5 * time.Minute
```
This is unfortunately way too long, e.g. we have 100-node scalability test clusters.
Can we perhaps pipeline/parallelise a few reboots at a time?  At least 2 so that we can make progress if one gets delayed.

quinton-hoole replied:
In retrospect I see that you're using a MIG rolling update, which is probably serial by definintion.  In which case you might need to limit the number of nodes that you reboot or something.  Taking 100 x 5 min = 8.3 hours to fail is clearly not tenable.  I'm out of ideas as to how to get this down to a few minutes, but we'll need to think of something.

mbforbes replied:
Check out this line and the comment I wrote above it:
https://github.com/GoogleCloudPlatform/kubernetes/pull/8837/files#diff-bbc40d90a96515627afa5efea9599dbeR292

We can just set this to the cluster size if we want it to run in parallel. Then, the whole thing should take roughly a single node's time.

I originally wrote it this way because during node upgrades, we expect to do it serially, so that if something fails, we destroy the minimum number of nodes we can :-)

However, for testing, we could either have a flag to do it serially for "normal" e2es and in parallel for scalability runs, or just always do it in parallel as it's probably going to be roughly the same test.

Another thought: for "soak" tests, should we be disabling all reboot / restart / recreate tests? So that we see how long the system really lasts?

quinton-hoole replied:
Yes, I agree.   I think the best we can do is:

1. Be default, only restart two nodes, in serial (i.e. devs do this by default, as do standard Jenkins runs)
2. In addition, we can try restarting all nodes in parallel, but I'm guessing that this is unlikely to be successful, and it's fairly unrepresentative of reality anyway.  Make it possible to do this via appropriate environment variables, and we can dedicate a Jenkins job to this.
3. Make it possible to run this test completely (i.e. restart every node, and in serial) using environment variables, and dedicate a Jenkins test run to this. This is the most realistic test, but unfortunately it takes too long to have it delay the standard e2e test runs.

quinton-hoole replied:
Oh, and regarding soak tests, yes, I think that all node restarts should be disabled there.

mbforbes replied:
1. By default, devs and jenkins launch clusters with only two nodes (a la [`cluster/gce/config-test.sh`](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/gce/config-test.sh#L23)), so this will happen. Furthermore, configuring to only rolling-update to some of the nodes would involve passing an `--auto-pause-after-instances` flag, which would change the test: we would need to look for "PAUSED" state instead of "ROLLED_OUT", but it's unclear how "PAUSED" is different than "CANCELLED", and which one will be applied if the rolling update fails. (I can't easily tell because I haven't gotten one to fail yet.) Anyway, tl;dr is we have two-node clusters, so this should only apply to two nodes.

2. I vote we punt on parallel right now. We can parameterize this and set up a dedicated Jenkins job if we decide to investigate this later. (I'm curious: why is it unlikely to be successful? If I understand correctly, all necessary state to reschedule all pods is on the master, so blowing away all nodes at once shouldn't cause any problems, right?)

3. Shouldn't be necessary (a la 1).

Re: soak tests, the only one I saw that disabled other tests (like reboot) was "kubernetes-soak-continuous-e2e", so I disabled this test on that already. Are there other Jenkins jobs I should do this to?

Re: large clusters, are there any Jenkins jobs that run > 2 size clusters for which I should also disable this test?

quinton-hoole replied:
1. That's only the default, which anyone can override.  I guess what we're saying is that anyone who overrides that default should also disable this test, or suffer horribly-long test run times.  I guess that's OK, although it's guaranteed to trip some people up.

2. Ack

3. OK, as per 1.

Re disabling on Jenkins jobs, yes, you'll want to disable it on all of the parallel runs, and the scalability runs.

quinton-hoole commented on test/e2e/restart.go:
```diff
@@ -0,0 +1,367 @@
+
+	// How long a node is allowed to become "Ready" after it is restarted before
+	// the test is considered failed.
+	restartNodeReadyAgainTimeout = 5 * time.Minute
```
It seems that this needs to be shorter than restartPerNodeTimeout to be useful?

mbforbes replied:
All of these timeouts come serially—this allows for 5 minutes AFTER the node is restarted (from an infrastructure point of view) before it is ready (from kubernetes' point of view).

quinton-hoole replied:
10 minutes is too long to wait for a failure in general.  But I can shrink it down in a separate PR once we know how long it usually takes in the successful case.

quinton-hoole commented on test/e2e/restart.go:
```diff
@@ -0,0 +1,367 @@
+
+	// How long a pod is allowed to become "running" and "ready" after a node
+	// restart before test is considered failed.
+	restartPodReadyAgainTimeout = 5 * time.Minute
```
As above.

mbforbes replied:
As above—this is AFTER the above timeout.

quinton-hoole replied:
ack.

quinton-hoole commented on test/e2e/restart.go:
```diff
@@ -0,0 +1,367 @@
+		// check must be identical to that call.
+		provider := testContext.Provider
+		nn := testContext.CloudConfig.NumNodes
+		if !providerIs("gce") {
```
@justinsb FYI . It would be good to extend this for AWS also.

quinton-hoole commented on test/e2e/restart.go:
```diff
@@ -0,0 +1,367 @@
+	var nodeList *api.NodeList
+	var err error
+	found, start := false, time.Now()
+	for ; time.Since(start) < nt; time.Sleep(poll) {
```
Please use the standard wait.Poll() method for retrying.

mbforbes replied:
Done (not yet pushed).

quinton-hoole commented on test/e2e/restart.go:
```diff
@@ -0,0 +1,367 @@
+			expect, nt, err)
+	}
+
+	// Next, ensure in parallel that all the nodes are ready. We subtract the
```
I think this warrants it's own function.  This one is getting a bit long.

mbforbes replied:
Ugh, yeah, I spent a while trying to figure out the best way for this to be factored... all this really does is run _n_ `waitForNodeToBeReady(...)` in parallel and check their result. However, because of our Go 1.3 support and my extra-safe "waiting until all nodes are done before failing," this a handful extra lines of code. I din't factor it out more because I wouldn't reuse the function if I did.

Please let me know your final decision after my whining above :-)

quinton-hoole replied:
Whining accepted :-)  But I think it's worth factoring out, even if the resulting function is never re-used.  It makes the code more readable.

mbforbes replied:
Ack.

quinton-hoole commented on test/e2e/restart.go:
```diff
@@ -0,0 +1,367 @@
+func migRollingUpdateStart(templ string, nt time.Duration) (string, error) {
+	var err error
+	prefix, suffix := "Started [", "]."
+	for i := 0; i < singleCallAttempts; i++ {
```
Please use wait.Poll() for retries here.

mbforbes replied:
Question: I'm doing something different for retries on single API calls versus "timeout"/"waiting" calls, where a single call is tried n times. This seemed slightly better choosing timeouts that all single calls are supposed to conform to. Do we really want to switch all of these to waiting for time periods? If so, what's a good periods? Single API calls *should* fail very quickly, so we could get all 5 failures here in ~2-3 seconds, but might that be too risky? If we have it be 20 seconds or 1 minute, we might call a few hundred times; is that what we want?

quinton-hoole replied:
I don't think that retrying something a fixed number of times as fast as possible is useful. If the server is overloaded, you just overload it even further, and fail anyway.
I'd suggest telling wait.Poll() to try every, say 5 seconds, for a maximum of 20 seconds.  That way you get about 3 retries, and a reasonable maximum elapsed time.  Sound good?

mbforbes replied:
Done (not yet pushed).

quinton-hoole commented on test/e2e/util.go:
```diff
@@ -178,7 +193,7 @@ func waitForPodsRunningReady(ns string, minPods int, timeout time.Duration) erro
 	Logf("Waiting up to %v for all pods (need at least %d) in namespace '%s' to be running and ready",
 		timeout, minPods, ns)
-	for start := time.Now(); time.Since(start) < timeout; time.Sleep(podPoll) {
+	for start := time.Now(); time.Since(start) < timeout; time.Sleep(poll) {
```
Please use wait.Poll() instead.

mbforbes replied:
Agh, the correct way to wait in e2es keeps changing under me! Will change this :-)

mbforbes replied:
Done (not yet pushed).

quinton-hoole commented on test/e2e/util.go:
```diff
@@ -204,20 +219,23 @@ func waitForPodsRunningReady(ns string, minPods int, timeout time.Duration) erro
-func waitForPodCondition(c *client.Client, ns, podName, desc string, poll, timeout time.Duration, condition podCondition) error {
-	Logf("Waiting up to %v for pod %s status to be %s", timeout, podName, desc)
+func waitForPodCondition(c *client.Client, ns, podName, desc string, timeout time.Duration, condition podCondition) error {
+	Logf("Waiting up to %[1]v for pod %-[2]*[3]s status to be %[4]s", timeout, podPrintWidth, podName, desc)
```
Please use wait.Poll() below.

mbforbes replied:
Done (not yet pushed).

quinton-hoole commented on test/e2e/util.go:
```diff
@@ -839,11 +857,22 @@ func DeleteRC(c *client.Client, ns, name string) error {
+// Convenient wrapper around listing nodes supporting retries.
+func listNodes(c *client.Client, label labels.Selector, field fields.Selector) (*api.NodeList, error) {
+	nodes, err := c.Nodes().List(label, field)
+	for i := 1; i < singleCallAttempts; i++ {
```
Please use wait.Poll()

mbforbes replied:
Done (not yet pushed).

quinton-hoole commented on test/e2e/util.go:
```diff
@@ -839,11 +857,22 @@ func DeleteRC(c *client.Client, ns, name string) error {
-	maxRetries := 4
 	pods, err := c.Pods(namespace).List(label, field)
-	for i := 0; i < maxRetries; i++ {
+	for i := 1; i < singleCallAttempts; i++ {
```
Please use wait.Poll()

mbforbes replied:
Done (not yet pushed).

quinton-hoole commented on test/e2e/util.go:
```diff
@@ -1025,6 +1054,74 @@ func getSigner(provider string) (ssh.Signer, error) {
+// ready or unknown).
+func waitForNodeToBe(c *client.Client, name string, wantReady bool, timeout time.Duration) bool {
+	Logf("Waiting up to %v for node %s readiness to be %t", timeout, name, wantReady)
+	for start := time.Now(); time.Since(start) < timeout; time.Sleep(poll) {
```
Please use wait.Poll()

mbforbes replied:
Done (not yet pushed).

quinton-hoole commented:
Reassigned to mforbes until this is ready for final review.

mbforbes commented:
Thanks for the review @quinton-hoole—I've responded with a couple questions (reassigning until you answer them, ha!). Just doing this so I can hopefully do one round of "addressing comments" before this LGTY.

Please ping/reassign once you've had a chance to answer... I'm ostensibly out today but would really like to get this in soon, so I can probably address them tonight.

mbforbes commented:
(Whoops, assigned to Robby out of instinct. Back to Quinton ... :-)

quinton-hoole commented:
OK, I'll do my best to finish this off today.  Got a few fires to fight first though :-)

mbforbes commented:
Ack! :fire_engine: :potable_water: :fire:

mbforbes commented:
Ping (for Monday) :-)

quinton-hoole commented:
OK, I think we're nearly done here.  Reassign to me when you're done.  I'll be OOTO until Wed 6/3 (back on Thu).

mbforbes commented:
Addressed comments, PTAL.

mbforbes commented:
Ping!

quinton-hoole commented:
I'm swamped under oncall right now.  I hope to get to this today.

mbforbes commented:
No biggie—feel free to hold off for Monday; probably not best to merge a new test on a Friday eve.

quinton-hoole commented:
LGTM

krousey commented:
@quinton-hoole This seems to introduce a new e2e test. Are there any concerns about the filters on our CI projects? Do restarts play well with other tests (both serially and in parallel)?

quinton-hoole commented:
Yes, as per my review comment "Re disabling on Jenkins jobs, yes, you'll want to disable it on all of the parallel runs, and the scalability runs.".  I'll do that now for you.  You can merge in the mean time.

mbforbes commented:
I did all of the test disabling last night after reading that comment—I hope there weren't any I missed?
